sympy/tensor/tensor.py

```diff
@@ -4063,26 +3980,9 @@ def perm2tensor(t, g, is_canon_bp=False):
     raise NotImplementedError()
 
 
+@deprecated(useinstead=".substitute_indices() instance method", issue=15276,
+            deprecated_since_version="1.5")
```

Comment:
Well, `.substitute_indices( )` will not work if you don't have a `TensExpr` object.

That's the reason why we have `substitute_indices( )`

Reply:
Sometimes you have scalar expression and you cannot determine which ones are. That's why we need this second syntax.

Reply:
I could not come up with a sensible example, so I decided to deprecate it. Can you come up with one where this function is useful, so that I keep it with a clean conscience?

Reply:
Suppose you want to apply `.substitute_indices` to the Taylor expansion of the exponential map of a tensor expression.

```
x*exp(A_i A^i) = x + x*A_i A^i + ...
```
The first term is not a subclass of `TensExpr`, so it will have no `.substitute_indices`.

Suppose you have two tensor additions:
```
expr1 = mu + A_i A^i
expr2 = theta + A_i A^i
```
now, `expr1 - expr2` may not be a tensor expression, therefore not having `.substitute_indices( ... )`.

Reply:
Right, but in both these examples there are no actual indices to substitute. All of them are dummy and are assigned a name automatically, so the function has really no purpose there.

The reason I am against having a static `substitute_indices()` method is that by the same logic you can expect to have plentiful others static methods, something like `get_free_indices()`, `get_dummy_indices()`, `get_data()`, `get_tensor_head()` and so on. Is it really necessary to have all of them? Or why implement ones and not the others?

Reply:
Combinations of tensor expressions may be cast to non-tensor expressions in certain cases.

The problem is that under some combination you may have these two conditions:
- you cannot deterministically determine the indices of an expression
- the expression may have been cast to a non-tensorial expression

The alternative would be to add 
```python
def substitute_indices(self, *args):
    return self
```
in SymPy core... so that all SymPy objects have `.substitute_indices`. But we don't want to define too many methods in SymPy core, because 99% of the users don't really need `.substitute_indices( )`.

By the way, this kind of API is pretty common in SymPy (e.g. `simplify( )` and `.simplify( )`).

Reply:
This is actually linked to a much more complex (still unsolved) problem in SymPy: how to add features to SymPy core objects without modifying SymPy's core.

In Wolfram Mathematica this problem is solved with pattern matching. We don't really have something similar here.

Reply:
Ok, I agree, there is some value in a static `substitute_indices()` method. I'll make the relevant corrections to the PR. Thank you for the comments!